lzy/portal/src/main/java/ai/lzy/portal/slots/StdoutSlot.java

```diff
@@ -105,55 +114,52 @@ public synchronized LzyInputSlot find(String slot) {
     }
 
     @Override
-    public synchronized Stream<ByteString> readFromPosition(long offset) {
-        assert offset == 0;
+    public void readFromPosition(long offset, StreamObserver<LSA.SlotDataChunk> responseObserver) {
+        assert offset == 0;  // This is stream slot
 
-        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(new Iterator<>() {
-            private ByteString line = null;
+        while (true) {
 
-            @Override
-            public boolean hasNext() {
-                synchronized (StdoutSlot.this) {
-                    while (true) {
-                        var l = buffer.poll();
-                        if (l != null) {
-                            line = ByteString.copyFromUtf8(l);
-                            return true;
-                        }
-
-                        task2slot.values().forEach(slot -> LOG.debug(" ::: slot {} -> {}", slot.name(), slot.state()));
-
-                        if (finished.get()) {
-                            return false;
-                        }
-
-                        try {
-                            StdoutSlot.this.wait();
-                        } catch (InterruptedException e) {
-                            // ignored
-                        }
-                    }
-                }
+            if (((ServerCallStreamObserver<LSA.SlotDataChunk>) responseObserver).isCancelled()) {
+                LOG.error("Stream cancelled, returning...");
+                return;
             }
 
-            @Override
-            public ByteString next() {
-                if (line == null) {
-                    throw new IllegalStateException();
+            var l = buffer.peek();
+            while (l != null) {
+                if (((ServerCallStreamObserver<LSA.SlotDataChunk>) responseObserver).isCancelled()) {
+                    LOG.error("Stream cancelled, returning...");
+                    return;
                 }
                 try {
-                    LOG.debug("Send from slot {} some data: {}", name(), line.toStringUtf8());
-                    return line;
-                } finally {
-                    try {
-                        onChunk(line);
-                    } catch (Exception re) {
-                        LOG.warn("Error in traffic tracker", re);
-                    }
-                    line = null;
+                    responseObserver.onNext(LSA.SlotDataChunk.newBuilder()
+                        .setChunk(ByteString.copyFromUtf8(l))
+                        .build());
+                    buffer.poll();
+                } catch (Exception e) {
+                    LOG.error("Error while sending data from slot {}", name(), e);
+                    responseObserver.onError(Status.INTERNAL.asException());
```

Comment:
it is useless in case of broken connection
we should check error more carefully